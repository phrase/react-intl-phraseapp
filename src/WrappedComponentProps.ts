import { WrappedComponentProps as ReactIntlWrappedComponentProps } from 'react-intl';

export type WrappedComponentProps = ReactIntlWrappedComponentProps;
