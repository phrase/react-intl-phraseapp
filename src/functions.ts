let phraseAppEditor = false;

function sanitizeConfig(config: any) : any {
  config.prefix = config.prefix ? config.prefix : '{{__';
  config.suffix = config.suffix ? config.suffix : '__}}';

  return config;
}

export function initializePhraseAppEditor (config: any) {
  if (phraseAppEditor) return;
  
  phraseAppEditor = true;
  (<any>window).PHRASEAPP_ENABLED = config.phraseEnabled;  
  (<any>window).PHRASEAPP_CONFIG = sanitizeConfig(config);

  if (config.phraseEnabled) {
      const phraseapp = document.createElement('script');
      phraseapp.type = 'text/javascript';
      phraseapp.async = true;
      phraseapp.src = ['https://', 'phraseapp.com/assets/in-context-editor/2.0/app.js?', new Date().getTime()].join('');
      var s = document.getElementsByTagName('script')[0];
      if (s != undefined) {
        s.parentNode.insertBefore(phraseapp, s);
      } else {
        document.insertBefore(phraseapp, null);
      }
    }
}

export function isPhraseEnabled() : boolean {
  return (<any>window).PHRASEAPP_ENABLED
}

export function escapeId (id : string) : string {
  let config = (<any>window).PHRASEAPP_CONFIG;
  if (config.removePhrasePrefix) {
    return  config.prefix + id + config.suffix;
  } else {
    return  config.prefix + 'phrase_' + id + config.suffix;
  }
}