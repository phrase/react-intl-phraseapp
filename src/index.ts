export { initializePhraseAppEditor } from './functions';
export { injectIntl, ReactIntlPhraseProps } from './injectIntl';
export { FormattedMessage } from './FormattedMessage';
export { useIntl } from './useIntl';
