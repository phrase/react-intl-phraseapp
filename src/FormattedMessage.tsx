import * as React from 'react';
import { FormattedMessage as IntlFormattedMessage } from 'react-intl';
import { Props } from 'react-intl/lib/src/components/message';

import { escapeId, isPhraseEnabled } from './functions';

export const FormattedMessage: React.FC<Props> = (props) => {
  if (isPhraseEnabled()) {
    return <span>{escapeId(props.id ?? '')}</span>;
  } else {
    return <IntlFormattedMessage {...props} />;
  }
};
