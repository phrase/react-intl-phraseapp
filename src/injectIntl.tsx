import * as React from 'react';
import { injectIntl as injectIntlReact, IntlFormatters, MessageDescriptor } from 'react-intl';
import { escapeId, isPhraseEnabled } from './functions';

export type ReactIntlPhraseProps = {
    translate: (_: string)=> string;
    formatMessage: IntlFormatters<any>['formatMessage'],
};

export function injectIntl(WrappedComponent: React.ComponentType<ReactIntlPhraseProps>, options?: Parameters<typeof injectIntlReact>[1]): ReturnType<typeof injectIntlReact> & React.FC<ReactIntlPhraseProps> {
    class InjectPhrase extends React.Component implements ReactIntlPhraseProps {
        constructor(...args: ConstructorParameters<typeof React.Component>) {
            super(...args);

            this.render = this.render.bind(this);
            this.translate = this.translate.bind(this);
            this.formatMessage = this.formatMessage.bind(this);
        }

        translate(keyName: string): ReturnType<ReactIntlPhraseProps['translate']> {
            if (isPhraseEnabled()) {
                const escapedString = keyName.replace("<", "[[[[[[html_open]]]]]]").replace(">", "[[[[[[html_close]]]]]]");
                return escapeId(escapedString);
            } else {
                return this.props['intl'].formatMessage({ "id": keyName });
            }
        }

        formatMessage(messageDescriptor: MessageDescriptor, _values?: any, _opts?: any): ReturnType<ReactIntlPhraseProps['formatMessage']> {
            const { id } = messageDescriptor;
            if (!id) {
                console.error("formatMessage requires an id")
            } else {
                return this.translate(id);
            }
        }

        render() {
            return (
                <WrappedComponent
                    translate={this.translate}
                    formatMessage={this.formatMessage}
                    {...this.props}
                />
            );
        }
    }

    return injectIntlReact(InjectPhrase, options) as any;
}
