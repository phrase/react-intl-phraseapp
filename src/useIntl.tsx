import * as React from 'react';
import { IntlShape, useIntl as ReactUseIntl } from 'react-intl';
import { escapeId, isPhraseEnabled } from './functions';

export function useIntl(): IntlShape {
  if (isPhraseEnabled()) {
    const intl = ReactUseIntl();
    function wrappedFormatMessage(...args) {
      return escapeId(args[0]?.id || '');
    }
    return { ...intl, formatMessage: wrappedFormatMessage };
  } else {
    return ReactUseIntl();
  }
};
