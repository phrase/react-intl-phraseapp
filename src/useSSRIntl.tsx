import { createIntl, createIntlCache } from '@formatjs/intl';
import { escapeId, isPhraseEnabled } from './functions';

export async function useSSRIntl({locale, messages}) {
  const cache = createIntlCache();
  const intl = createIntl({locale, messages}, cache)

  if (isPhraseEnabled()) {
    function wrappedFormatMessage(...args) {
      return escapeId(args[0]?.id || '');
    }
    return { ...intl, formatMessage: wrappedFormatMessage };
  } else {
    return intl;
  }
};
