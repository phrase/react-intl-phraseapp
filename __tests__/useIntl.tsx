import * as React from 'react';
import * as Utils from '../helper/test_utils';
import { injectIntl, useIntl } from '../src';

type MessageDescriptor = {
  id: string;
  defaultMessage?: string;
  description?: string | object;
};

let ComponentUnderTest;
const messages: Record<string, MessageDescriptor> = {
  greeting: {
    id: 'app.greeting',
    defaultMessage: 'Hello, {name}!',
    description: 'Greeting to welcome the user to the app',
  },
};

describe('useIntl', () => {
  beforeEach(() => {
    function Component() {
      const { formatMessage } = useIntl();
      return (<div>{formatMessage(messages.greeting, { name: 'Eric' })}</div>);
    }
    ComponentUnderTest = Component;
  });

  describe('useIntl', () => {
    it('used underlying useIntl', () => {
      const { container } = Utils.createComponentWithIntl(<ComponentUnderTest />);
      expect(container.firstChild).toMatchSnapshot();
    });
  });

  describe('with phrase enabled', () => {
    beforeEach(() => {
      Utils.setPhraseConfig();
    });

    it('uses wrapped useIntl', () => {
      const { container } = Utils.createComponentWithIntl(<ComponentUnderTest />);
      expect(container.firstChild).toMatchSnapshot();
    });
  });
});
