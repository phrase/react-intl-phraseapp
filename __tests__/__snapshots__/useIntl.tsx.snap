// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`useIntl useIntl used underlying useIntl 1`] = `
<div>
  Hello, Eric!
</div>
`;

exports[`useIntl with phrase enabled uses wrapped useIntl 1`] = `
<div>
  {{__phrase_app.greeting__}}
</div>
`;
